kostyor/upgrades/engine.py

```diff
@@ -152,17 +152,21 @@ def start(self):
         # keystone, then with nova, and so on. See iteration details in
         # get_controllers() docstring.
         for host in get_controllers(hosts):
+            self.driver.pre_host_upgrade_hook(host)
             for service in iterservices(host):
+                self.driver.pre_service_upgrade_hook(service)
```

Comment:
There are cons/pros in both approaches. I think my concern is having a lot of methods in the interface leads to confusing and complexity in maintaining. I like when there's only one way to do a thing. Your proposal allows both approaches.